lib/ecto/query/builder.ex

```diff
@@ -791,6 +791,84 @@ defmodule Ecto.Query.Builder do
   defp do_literal(value, expected, _current),
     do: {:%, [], [Ecto.Query.Tagged, {:%{}, [], [value: value, type: expected]}]}
 
+  @doc """
+  Escape a values list and extract its parameters
+  """
+  @spec escape_values_list(Macro.t(), Macro.t()) :: {Macro.t(), Macro.t()}
+  def escape_values_list(values_list, types) do
+    fields =
+      quote do
+        Ecto.Query.Builder.values_list_fields(unquote(values_list))
+      end
+
+    types =
+      quote do
+        Ecto.Query.Builder.values_list_types!(unquote(fields), unquote(types))
+      end
+
+    params =
+      quote do
+        Ecto.Query.Builder.values_list_params!(unquote(values_list), unquote(types))
+      end
+
+    num_rows = quote do: length(unquote(values_list))
+
+    {{:{}, [], [:values, [], [types, num_rows]]}, params}
+  end
+
+  @doc """
+  Extract the field names from a values list
+  """
+  @spec values_list_fields(list(map)) :: map()
+  def values_list_fields(values_list) do
+    fields =
+      Enum.reduce(values_list, %{}, fn values, fields ->
+        Enum.reduce(values, fields, fn {field, _}, fields ->
+          Map.put(fields, field, true)
+        end)
+      end)
+
+    Map.keys(fields)
```

Comment:
```suggestion
          Map.put(fields, field, true)
        end)
      end)

    Map.keys(fields)
```
This is a MapSet, so you could use it for simplicity

Reply:
done! thank you